nucypher/utilities/oracles.py

```diff
@@ -0,0 +1,120 @@
+ This file is part of nucypher.
+
+ nucypher is free software: you can redistribute it and/or modify
+ it under the terms of the GNU Affero General Public License as published by
+ the Free Software Foundation, either version 3 of the License, or
+ (at your option) any later version.
+
+ nucypher is distributed in the hope that it will be useful,
+ but WITHOUT ANY WARRANTY; without even the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ GNU Affero General Public License for more details.
+
+ You should have received a copy of the GNU Affero General Public License
+ along with nucypher.  If not, see <https://www.gnu.org/licenses/>.
+"""
+from abc import ABC, abstractmethod
+from typing import Optional
+
+import requests
+from web3 import Web3
+from web3.gas_strategies.rpc import rpc_gas_price_strategy
+from web3.types import Wei, TxParams
+
+
+class Oracle(ABC):
+
+    class OracleError(RuntimeError):
+        """Base class for Oracle-related exceptions"""
+
+    name = NotImplemented
+    api_url = NotImplemented  # TODO: Deal with API keys
+
+    def _probe_oracle(self):
+        try:
+            response = requests.get(self.api_url)
+        except requests.exceptions.ConnectionError as e:
+            error = f"Failed to probe oracle at {self.api_url}: {str(e)}"
+            raise self.OracleError(error)
+
+        if response.status_code != 200:
+            error = f"Failed to probe oracle at {self.api_url} with status code {response.status_code}"
+            raise self.OracleError(error)
+
+        self._raw_data = response.json()
+
+    def __repr__(self):
+        return f"{self.name} ({self.api_url})"
+
+
+class EthereumGasPriceOracle(Oracle):
+    """Base class for Ethereum gas price oracles"""
+
+    _speed_names = NotImplemented
+    _default_speed = NotImplemented
+
+    @abstractmethod
+    def _parse_gas_prices(self):
+        return NotImplementedError
+
+    def get_gas_price(self, speed: Optional[str] = None) -> Wei:
+        speed = speed or self._default_speed
+        self._parse_gas_prices()
+        gas_price_wei = Wei(self.gas_prices[speed])
+        return gas_price_wei
+
+    @classmethod
+    def construct_gas_strategy(cls):
+        def oracle_based_gas_price_strategy(web3: Web3, transaction_params: TxParams = None) -> Wei:
+            oracle = cls()
+            gas_price = oracle.get_gas_price()
+            return gas_price
+        return oracle_based_gas_price_strategy
+
+
+class EtherchainGasPriceOracle(EthereumGasPriceOracle):
+    """Gas price oracle from Etherchain"""
+
+    name = "Etherchain oracle"
+    api_url = "https://www.etherchain.org/api/gasPriceOracle"
+    _speed_names = ('safeLow', 'standard', 'fast', 'fastest')
+    _default_speed = 'fast'
+
+    def _parse_gas_prices(self):
+        self._probe_oracle()
+        self.gas_prices = {k: int(Web3.toWei(v, 'gwei')) for k, v in self._raw_data.items()}
+
+
+class UpvestGasPriceOracle(EthereumGasPriceOracle):
```

Comment:
I don't really see a reason for this design pattern to be separated per API.

It makes more sense to me that you would have a single gas price oracle which fetches from multiple sources. Why not just combine these two classes into one gas price oracle?

Reply:
Seems like upvest has a more sophisticated feed that handles price spikes and source degradation. @cygnusv is one of these the default oracle to get a price more cheaply? Are there compatibility differences with Infura vs Alchemy?

Reply:
This PR creates the possibility of using this data feed, but it requires more work to make it more customizable. Right now, what it does is:
- When using a normal provider, the default is to use the "fast" web3py strategy. You can still configure it manually to use a different web3py strategy (but not the ones from the data feed)
- When using Infura, the default is to use a chain of data feeds, starting from Etherchain, then Upvest, and then the default of the provider (which relies on their own ETH nodes). In this case, you can't change the strategy manually.

So, this PR is somewhat very opinionated on the matter, but basically because I didn't want to spend too much time building a truly generic and customizable mechanism. We can do that in a follow-up PR.